### PR TITLE
Introduce `add_named_filter` to `OriginTrait`

### PR DESCRIPTION
**Pull request by bkchr**
The `add_named_filter` can be used in contexts where you have filters that depend on call local data. When you then have nested calls, you may only want that the context of the latest call in the stack is respected for in the call filters. An example are nested proxy calls with different proxy types where the inner call wants to add a new proxy.

### CODE REVIEW

**Review by ruseinov — APPROVED** [review]
lgtm, there are a bunch of old failing checks that confused me at first.

File: frame/support/procedural/src/construct_runtime/expand/origin.rs [hunk 1]
```diff
@@ -138,9 +145,20 @@ pub fn expand_outer_origin(
...
+				});
+			}
+
+			fn add_named_filter<Pallet: #scrate::traits::PalletInfoAccess, F: Fn(&Self::Call) -> bool + 'static>(
+				&mut self,
+				name: &'static str,
+				filter: F,
+			) {
+				self.named_filters.insert(
+					(<Pallet as #scrate::traits::PalletInfoAccess>::name(), name),
```
Comment (ggwpez) [inline frame/support/procedural/src/construct_runtime/expand/origin.rs:159]: Does this work if there are multiple instances of the same pallet?

**Comment by ggwpez** [comment]
bot rebase

File: frame/support/procedural/src/construct_runtime/expand/origin.rs (hunk 1 above)
Comment (bkchr) [reply frame/support/procedural/src/construct_runtime/expand/origin.rs:159]: Yes, each instance has its own name in the runtime

File: frame/support/procedural/src/construct_runtime/expand/origin.rs [hunk 2]
```diff
@@ -160,7 +179,7 @@ pub fn expand_outer_origin(
 				match self.caller {
 					// Root bypasses all filters
 					OriginCaller::system(#system_path::Origin::<#runtime>::Root) => true,
-					_ => (self.filter)(call),
+					_ => self.named_filters.values().all(|f| (*f)(call)) && (*self.filter)(call),
```
Comment (shawntabrizi) [inline frame/support/procedural/src/construct_runtime/expand/origin.rs:182]: i assume it is not really possible to overload the number of filters, making this logic heavy?

File: frame/proxy/src/lib.rs [hunk 3]
```diff
@@ -805,11 +804,11 @@ impl<T: Config> Pallet<T> {
 
 	fn do_proxy(
 		def: ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>,
+		mut origin: OriginFor<T>,
 		real: T::AccountId,
 		call: <T as Config>::Call,
 	) {
-		// This is a freshly authenticated new account, the origin restrictions doesn't apply.
-		let mut origin: T::Origin = frame_system::RawOrigin::Signed(real).into();
```
Comment (shawntabrizi) [inline frame/proxy/src/lib.rs:812]: hmm... maybe we should double check with whoever wrote this comment.

**Review by shawntabrizi — APPROVED** [review]

File: frame/proxy/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/proxy/src/lib.rs:812]: although it doesn't make sense to me now

**Comment by bkchr** [comment]
After some discussion we came to the conclusion that this actually isn't needed.

**Pull request closed without merging**